### PR TITLE
Interactive configuration

### PR DESCRIPTION
**Pull request by lstein**
# Add an interactive configuration step at install time

- After running the install script, the install script will call `configure_invokeai.py`. This is a renamed version of `preload_models.py`. The latter script is still present, but as a thin backward-compatibility shell.
- At install time, if no runtime root directory exists and no `.invokeai` startup file is present, the user will be prompted to select the runtime and outputs directories.
-  User can change the location of the root directory by editing `~/.invokeai`, specifying `--root` when calling `invoke.py`, or by setting the environment variable `INVOKEAI_ROOT`.
- This PR also expand the number of initial models offered to the user to include the  most "liked" ones from HuggingFace, including the two trinart models, the  PaperCut model, and the VoxelArt model (see below).
- Introduced an installation configuration file called `configs/INITIAL_MODELS.yaml` to make adding new models to the install script easier.
Screenshot from 2022-11-19 14-31-09

### CODE REVIEW

**Comment by hipsterusername** [comment]
We may want to make specific init options more explicit, either in-line after a command or by linking out.

How are we making determinations on included models?

**Comment by tildebyte** [comment]
Generally LGTM, but I don't really get what's going on with the completer...

**Review by Kyle0654 — APPROVED** [review]

**Comment by lstein** [comment]
> Generally LGTM, but I don't really get what's going on with the completer...

The completer needed a little work to autocomplete on directory paths outside the source code repo.

**Comment by lstein** [comment]
> We may want to make specific init options more explicit, either in-line after a command or by linking out.
> 
> 
> 
> How are we making determinations on included models?

I picked the top five hearted SD models on HuggingFace. There's a popular Chinese language version of SD there too ("Taiyu") but it requires a different tokenizer to work and there was too much surgery involved)

**Comment by damian0815** [comment]
i was requested to review but i don't know how to resolve the conflict

**Comment by tildebyte** [comment]
@lstein; I realized late yesterday that the installation per se should complete *before* we run load_models (specifically: we can end up in a situation where the invoke script hasn't been copied into place) - that's all that the conflict is; moving the call to load_models in both install scripts

File: installer/install.bat
Comment (tildebyte) [inline installer/install.bat:145]: These deactivate statements are **necessary**, so that the user isn't left in an active virtual environment after the *install* is done. Imagine they immediately launch the invoke script - now they're 2 levels deep.

File: installer/install.sh
Comment (tildebyte) [inline installer/install.sh:212]: Idem

File: scripts/configure_invokeai.py
Comment (psychedelicious) [inline scripts/configure_invokeai.py:581]: If `Globals.root` is `''` (empty string), and you accept the defaults, the end result is that, inside `initialize_rootdir()`, `root` and `outputs` are also empty strings. We need to return the sensible defaults.

```suggestion
#-------------------------------------
def select_root(yes_to_all:bool=False):
    default = os.path.expanduser('~/invokeai')
    if (yes_to_all):
        return default
    completer.set_default_dir(default)
    completer.complete_extensions(())
    completer.set_line(default)
    selected_directory = input(f"Select a directory in which to install InvokeAI's models and configuration files [{default}]: ")
    return selected_directory or default

#-------------------------------------
def select_outputs(root:str,yes_to_all:bool=False):
    default = os.path.normpath(os.path.join(root,'outputs'))
    if (yes_to_all):
        return default
    completer.set_default_dir(os.path.expanduser('~'))
    completer.complete_extensions(())
    completer.set_line(default)
    selected_directory = input(f'Select the default directory for image outputs [{default}]: ')
    return selected_directory or default
```

File: scripts/configure_invokeai.py
Comment (psychedelicious) [inline scripts/configure_invokeai.py:601]: The script crashes if the `outputs` exists. Needs `exist_ok=True`.

```suggestion
    os.makedirs(outputs, exist_ok=True)
```

**Review by psychedelicious — CHANGES_REQUESTED** [review]
One final change - not sure where to put this. 

We need to set the `PYTORCH_ENABLE_MPS_FALLBACK` environment variable to `1` on MPS devices. This is seems to not be done at the moment.

**Comment by damian0815** [comment]
developer console? that’s very windows nomenclature and as a mac user i wouldn’t expect to have to go there just to download a model file

can the root invoke.sh menu gain a „interactive configuration“ option?

**Comment by tildebyte** [comment]
@psychedelicious, @damian0815
 
> We need to set the `PYTORCH_ENABLE_MPS_FALLBACK` environment variable to `1` on MPS devices. This is seems to not be done at the moment.

I could've *sworn* I'd already added that to invoke.sh (which is the right place for it)...

I can add it to #1537 if there are no objections

**Comment by damian0815** [comment]
> I can add it to #1537 if there are no objections

yeah it'll need to be in install.sh as well i would say

to be clear though if a user has to re-run configure_invokeai.py they only way they can currently do it is by manually setting that environment variable.

like i suggested above, the best way to do this is to add a third option to invoke.sh that calls configure_invokeai.py. then invoke.sh could take a command line argument eg `web` `cli` `configure` or `devconsole` that bypasses the interactive step and just executes the requested script/mode.

**Comment by tildebyte** [comment]
RE:  `PYTORCH_ENABLE_MPS_FALLBACK`

> yeah it'll need to be in install.sh as well i would say
>
> to re-run configure_invokeai.py they only way they can currently do it is by manually setting that environment variable.

For my edification - why is this? What part does torch play in *running* the install or configure scripts?

**Comment by damian0815** [comment]
uh, maybe it doesn't? @psychedelicious why is `PYTORCH_ENABLE_MPS_FALLBACK=1` needed here? i know we need it to run invoke.py, why do we need it (do we need it?) to run `configure_invokeai.py`?

**Comment by psychedelicious** [comment]
> uh, maybe it doesn't? @psychedelicious why is PYTORCH_ENABLE_MPS_FALLBACK=1 needed here? i know we need it to run invoke.py, why do we need it (do we need it?) to run configure_invokeai.py?

Sorry, I wasn't clear. It needs to be set in the venv activation scripts so that when you run invoke.sh and choose option 3 you can actually run invoke.py manually.

E.g. I had to edit my `activate.fish`, adding `set PYTORCH_ENABLE_MPS_FALLBACK 1` (this is the `fish` syntax, doesn't use `export`) to then run `python scripts/invoke.py`. 

While helping user @whosawhatsis on discord, we had to do this. @tildebyte had helped him update his `install.sh` to point to the current `development.tar.gz` archive and working through the issues there prompted my requested changes on this PR.

File: installer/install.bat
Comment (lstein) [reply installer/install.bat:145]: I have put the deactivate statements back in.

File: scripts/configure_invokeai.py
Comment (lstein) [reply scripts/configure_invokeai.py:581]: Responses:
1. I've changed the logic a bit. When the user provides a relative path for root, including the empty string and '.'. The script converts the path into an abspath and configures that as the root. So installs to '' and '.' will both go into the repo, which is consistent with previous behavior.
2. Fixed
3. Fixed

**Comment by lstein** [comment]
> @psychedelicious, @damian0815
> 
> > We need to set the `PYTORCH_ENABLE_MPS_FALLBACK` environment variable to `1` on MPS devices. This is seems to not be done at the moment.
> 
> I could've _sworn_ I'd already added that to invoke.sh (which is the right place for it)...
> 
> I can add it to #1537 if there are no objections

This is a separate issue with the installer which I _think_ is addressed in recent installer fixes pull request. I'll make sure.

**Comment by lstein** [comment]
I believe this is ready to go. @psychedelicious I will need your approval on the fixes I made to address your change requests.

**Review by psychedelicious — CHANGES_REQUESTED** [review]
- The defaults still do not work. 

I removed my `.invokeai` folder before running `configure_invokeai.py`. You probably need to do this to recreate the issue. There's nowhere in the code that actually sets `Globals.root` to anything if it is empty to start with. 

Here is what happened:

When prompted for root and outputs directories, the expected sane defaults displayed but where not selected:

```
Select a directory in which to install InvokeAI's models and configuration files [/Users/spencer/invokeai]: # pressed enter
Select the default directory for image outputs [outputs]: # pressed enter
InvokeAI models and configuration files will be placed into  and image outputs will be placed into /Users/spencer/installer.
```

So when I proceeded to download models, it immediately started downloading them to the wrong place:
```
** DOWNLOADING WEIGHTS **
/Users/spencer/installer/models/ldm/stable-diffusion-v1/ # wrong directory!
* v1-5-pruned-emaonly.ckpt: Downloading...
````

- The format string is fixed.
- Now if I enter a path with an escaped space e.g `/Users/spencer/invoke\ ai/`, it creates that path exactly (with the slash in it). I think this is fine.

**Comment by lstein** [comment]
> * The defaults still do not work.
> 
> I removed my `.invokeai` folder before running `configure_invokeai.py`. You probably need to do this to recreate the issue. There's nowhere in the code that actually sets `Globals.root` to anything if it is empty to start with.
> 
> Here is what happened:
> 
> When prompted for root and outputs directories, the expected sane defaults displayed but were not selected:
> 
> ```
> Select a directory in which to install InvokeAI's models and configuration files [/Users/spencer/invokeai]: # pressed enter
> Select the default directory for image outputs [outputs]: # pressed enter
> InvokeAI models and configuration files will be placed into  and image outputs will be placed into /Users/spencer/installer.
> ```
I just realized that this is an issue with the Windows implementation of the readline library. On Mac and Linux, the correct default is placed in the readline buffer for editing, but Windows doesn't support this operation. I'll work around this.

**Comment by psychedelicious** [comment]
> I just realized that this is an issue with the Windows implementation of the readline library. On Mac and Linux, the correct default is placed in the readline buffer for editing, but Windows doesn't support this operation. I'll work around this.

I am on macOS, no Windows in this house

**Review by psychedelicious — APPROVED** [review]
Just tested, default paths working now. Thank you!

File: scripts/configure_invokeai.py
Comment (psychedelicious) [inline scripts/configure_invokeai.py:601]: Typo: `--output_dir` to `--outdir`

```suggestion
    print(f'\nYou may change the chosen directories at any time by editing the --root and --outdir options in "{Globals.initfile}",')
```

**Review by psychedelicious — CHANGES_REQUESTED** [review]

**Comment by lstein** [comment]
> > I just realized that this is an issue with the Windows implementation of the readline library. On Mac and Linux, the correct default is placed in the readline buffer for editing, but Windows doesn't support this operation. I'll work around this.
> 
> I am on macOS, no Windows in this house

Ooch. Maybe it only works on Linux! Anyway, the defaults are now fixed.

**Comment by lstein** [comment]
@psychedelicious I believe you have to change your review to "approve" before I can merge this. As far as I'm aware the only pending change was the documentation typo in `--outdir`, and this has been committed.

**Review by psychedelicious — APPROVED** [review]
LGTM

**Pull request merged**
